### PR TITLE
Updates and Fixes.

### PR DESCRIPTION
**Pull request by percymamedy**
- Move all config to event-bus.php config file.
- Add customization of event type, queue and connection and update README to reflect it.
- Handle token expiry when sending events.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -128,6 +117,169 @@ class ListingController extends Controller
...
+            'description' => $this->listing->description
+        ]);
+}
+```
+
+You may also implement a `broadcastToEventBusAs` method on your Laravel Event class, however note that `withEventType` will
+take precedence over `broadcastToEventBusAs`.
+
+```php
+<?php
```
Comment (datashaman) [inline README.md]: dont need this php header i think. we're not executing the readme. :)

File: src/EventBusServiceProvider.php [hunk 2]
```diff
@@ -38,12 +38,25 @@ public function register()
         $this->configure();
 
         $this->app->bind(Client::class, function () {
-            return new Client(config('services.service_bus'));
+            $clientConfig = [
+                'culture' => config('event-bus.culture'),
```
Comment (datashaman) [inline src/EventBusServiceProvider.php]: Structure the config the way it should be to use it. None of this code should be here, it was better before.

**Review by datashaman — CHANGES_REQUESTED** [review]

File: src/EventBusServiceProvider.php (hunk 2 above)
Comment (datashaman) [reply src/EventBusServiceProvider.php]:
```
return new Client(config('event-bus'));
```

File: src/Listeners/DispatchBroadcastToEventBusJob.php [hunk 3]
```diff
@@ -16,10 +16,27 @@ class DispatchBroadcastToEventBusJob
...
+
+        $queue = config('event-bus.queue');
+        $connection = config('event-bus.queue_connection');
+
+        if (method_exists($event, 'broadcastToEventBusOnQueue')) {
+            $queue = $event->broadcastToEventBusOnQueue($busEvent);
+        }
+
+        if (method_exists($event, 'broadcastToEventBusOnQueueConnection')) {
+            $connection = $event->broadcastToEventBusOnQueueConnection($busEvent);
```
Comment (datashaman) [inline src/Listeners/DispatchBroadcastToEventBusJob.php]: that is a _long_ method name. can these not be shorter?

Comment (datashaman) [reply src/Listeners/DispatchBroadcastToEventBusJob.php]: what about `onQueue` and `onConnection`?

File: src/Listeners/DispatchBroadcastToEventBusJob.php [hunk 4]
```diff
@@ -16,10 +16,27 @@ class DispatchBroadcastToEventBusJob
...
     public function handle(ShouldBroadcastToEventBus $event)
     {
-        $busEvent = $event->withServiceBusEventAs(
-            Event::make(class_basename($event))
-        );
+        $eventType = class_basename($event);
 
-        BroadcastToEventBus::dispatch($busEvent);
+        if (method_exists($event, 'broadcastToEventBusAs')) {
+            $eventType = $event->broadcastToEventBusAs();
```
Comment (datashaman) [inline src/Listeners/DispatchBroadcastToEventBusJob.php:22]: What about `eventType`?

File: src/Listeners/DispatchBroadcastToEventBusJob.php [hunk 5]
```diff
@@ -16,10 +16,27 @@ class DispatchBroadcastToEventBusJob
...
+        }
+
+        $busEvent = $event->withServiceBusEventAs(Event::make($eventType));
```
Comment (datashaman) [inline src/Listeners/DispatchBroadcastToEventBusJob.php]: What about `toEvent`?

Comment (datashaman) [reply src/Listeners/DispatchBroadcastToEventBusJob.php]: If the event is always made in this way, why do you have it as a parameter? Let the consumer make it inside the method themselves, I reckon.
```
$busEvent = $event->withServiceBusEventAs();
```
Then you don't need to have `broadcastToEventBusAs`, the consumer can decide what they want to call it in `toEvent`.

Comment (datashaman) [reply src/Listeners/DispatchBroadcastToEventBusJob.php]: Some example code with my naming suggestions:
```
public function handle(ShouldBroadcastToEventBus $event)
{
    $busEvent = $event->toEvent();

    $queue = method_exists($event, 'onQueue')
        ? $event->onQueue($busEvent)
        : config('event-bus.queue');

    $connection = method_exists($event, 'onConnection')
        ? $event->onConnection($busEvent)
        : config('event-bus.queue_connection');

    BroadcastToEventBus::dispatch($busEvent)
        ->onQueue($queue)
        ->onConnection($connection);
}
```

```
public function toEvent(): Event
{
    return Event::make('ListingCreatedEvent')
        ->withBlah();
}
```

Comment (datashaman) [reply src/Listeners/DispatchBroadcastToEventBusJob.php]: `toBusEvent` might be better, idk.

Comment (percymamedy) [reply src/Listeners/DispatchBroadcastToEventBusJob.php]: The consumer has the option to either implement the method "broadcastToEventBusAs" or use the Event from the parameter to customise the event type

```
public function withServiceBusEventAs(Event $event): Event
{
    return $event
        ->withEventType('UserListingCreatedEvent')
        ->withAction('user', $this->listing->user_id)
        ->withPayload([
            'id' => $this->listing->id,
            'title' => $this->listing->title,
            'description' => $this->listing->description
        ]);
}
```

And I want the event type to take the class name by default when they don't specify it. That's why I have to construct the event first and pass it to them as a parameter.

File: src/Listeners/DispatchBroadcastToEventBusJob.php (hunk 3 above)
Comment (assertchris) [reply src/Listeners/DispatchBroadcastToEventBusJob.php]: I like the long method names, but I also like Objective-C...

File: src/Client.php [hunk 6]
```diff
@@ -88,32 +98,45 @@ protected function request(): PendingRequest
...
+    /**
+     * Generate a new token for authentication with event bus.
+     *
+     * @return string
+     * @throws RequestException
+     */
+    protected function generateToken(): string
+    {
+        $response = $this->request()
+                         ->retry(3, 100)
```
Comment (assertchris) [inline src/Client.php:124]: Uncommon formatting. Surprised Style-CI doesn't complain.

**Review by assertchris — APPROVED** [review]

**Pull request merged**